### PR TITLE
Rework mempool sync logic.

### PR DESCRIPTION
**Pull request by gregorycollins**
- remove old subscription semantics, replace with a "recent inserts" log

  - change REST getPending API to optionally tail the recent inserts log starting from a "high water mark"

  - update current polling sync algorithm to do initial bi-directional sync, then perform subsequent sync using high water mark. Pull-only for now.

### CODE REVIEW

File: test/Chainweb/Test/Mempool.hs [hunk 1]
```diff
@@ -58,24 +53,39 @@ import qualified Numeric.AffineSpace as AF
...
+
+genNonEmpty :: PropertyM IO [MockTx]
+genNonEmpty = do
+    xs <- pick arbitrary
+    pre (not $ null xs)
+    return xs
+
+genTwoSets :: PropertyM IO ([MockTx], [MockTx])
+genTwoSets = do
+    xs <- (uniq . sortBy ord) <$> genNonEmpty
```
Comment (marklnichols) [inline test/Chainweb/Test/Mempool.hs]: You can remove the brackets in these two lines

**Comment by larskuhtz** [comment]
Before we merge, we may test it on testnet on a subset of the nodes, say, only euCommon.

**Comment by gregorycollins** [comment]
> Before we merge, we may test it on testnet on a subset of the nodes, say, only euCommon.

Has to be all or nothing, the mempool REST api is changing.

File: src/Chainweb/Mempool/Mempool.hs [hunk 2]
```diff
@@ -306,13 +305,13 @@ syncMempools' log0 us localMempool remoteMempool =
     deb :: Text -> IO ()
     deb = log0 Debug
 
-    sync = flip finally (deb "sync finished") $ do
-        deb "Get pending hashes from remote"
+    sync = flip finally (deb "sync exiting") (initialSync >>= subsequentSync)
```
[the remark this answers is not in the document]
Comment (gregorycollins) [reply src/Chainweb/Mempool/Mempool.hs]: Now that the first arg is short, no

**Pull request merged**